tensorboard/plugins/text/text_plugin.py

```diff
@@ -231,13 +293,19 @@ def index_impl(self):
       run_to_series[run] += tags.keys()
     return run_to_series
 
+  def tags_impl(self):
+    # Recompute the index on demand whenever tags are requested, but do it
+    # in a separate thread to avoid blocking.
+    self._maybe_launch_index_impl_thread()
```

Comment:
The projector plugin makes an assumption: Once it is deemed to be active, it never becomes inactive. Hence, you might consider assuming the same here - cache when the plugin is found to be active and never launch a separate thread again following that ... I think I'll defer to you to make the judgement here though.

Reply:
Yes, I actually opted to cache a little more aggressively than the projector plugin - the whole result of `index_impl()` is cached in `self._index_cached`, which has the benefit that `tags_impl()` can also return without blocking.  However, unlike `is_active()` we don't want `tags_impl()` to stop refreshing the cache once it's gotten a positive result, because even a non-empty index (one that corresponds to an active plugin) may not be complete if more text summary data is still being written out to the event files.  So `tags_impl()` needs to keep launching that background thread to refresh the cache as long as it's still being invoked itself.

Once we're at that point, we could either decide to separately cache the `is_active()` state permanently once it's determined to be True.  This would have the benefit that `is_active()` would be even cheaper once it's true.  However, I figured that it was slightly more complex to have two different pieces of cached data, and that we might as well just keep `is_active()` simple and have it reflect whatever the current active state is based on the last refresh of `_index_cached` - after all, if the plugin really doesn't ever go inactive after being active, the end result is the same.

Note that I do make it so that `is_active()` itself won't launch a new thread if the current state is active.  It's only via `tags_impl()` that a new thread will be kicked off unconditionally; `is_active()` will take advantage of fresher data if it's present but won't actually refresh the data itself once the plugin is deemed active.

Reply:
Ah, I see what you mean. We still need to launch the thread within `tags_impl` so that requests for tags stay fresh. SG.